lib/src/authentication/token_store.dart

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: import_of_legacy_library_into_null_safe
+
+import 'dart:convert';
+
+import 'package:path/path.dart' as path;
+
+import '../io.dart';
+import '../log.dart' as log;
+import '../source/hosted.dart';
+import 'token.dart';
+
+/// Stores and manages authentication credentials.
+class TokenStore {
+  TokenStore(this.cacheRootDir);
+
+  /// Cache directory.
+  final String cacheRootDir;
+
+  List<Token>? _tokens;
+
+  /// List of saved authentication tokens.
+  ///
+  /// Modifying this field will not write changes to the disk. You have to call
+  /// [flush] to save changes.
+  List<Token> get tokens => _tokens ??= _loadTokens();
+
+  /// Reads "tokens.json" and parses / deserializes it into list of
+  /// [Token].
+  List<Token> _loadTokens() {
+    final result = List<Token>.empty(growable: true);
+    final path = _tokensFile;
+    if (!fileExists(path)) {
+      return result;
+    }
+
+    try {
+      final json = jsonDecode(readTextFile(path));
```

Comment:
```suggestion
     try {
      final json = jsonDecode(readTextFile(path));
       } on FormatException {
          throw FormatException('$path is not valid JSON');
       }
```

This is probably better than a message saying: "did not expect ," :see_no_evil: